agent/src/util/discovery_operator.rs

```diff
@@ -0,0 +1,1394 @@
+        self.config.clone()
+    }
+    /// Returns instance_map field. Allows the struct to be mocked.
+    #[allow(dead_code)]
+    pub fn get_instance_map(&self) -> InstanceMap {
+        self.instance_map.clone()
+    }
+    /// Returns device_plugin_path field. Allows the struct to be mocked.
+    #[allow(dead_code)]
+    pub fn get_device_plugin_path(&self) -> String {
+        self.device_plugin_path.clone()
+    }
+    #[allow(dead_code)]
+    pub async fn stop_all_discovery(&self) {
+        let mut discovery_handler_map = self.discovery_handler_map.lock().unwrap().clone();
+        if let Some(protocol_dhs_map) =
+            discovery_handler_map.get_mut(&self.config.spec.protocol.name)
+        {
+            for (endpoint, dh_details) in protocol_dhs_map.clone() {
+                match dh_details.stop_discovery.send(()) {
+                    Ok(_) => trace!("stop_all_discovery - discovery client for protocol {} at endpoint {} told to stop", self.config.spec.protocol.name, endpoint),
+                    Err(e) => error!("stop_all_discovery - discovery client for protocol {} at endpoint {} could not receive stop message with error {:?}", self.config.spec.protocol.name, endpoint, e)
+                }
+            }
+        }
+    }
+
+    /// Calls discover on the Discovery Handler at the given endpoint and returns the connection stream.
+    pub async fn get_stream(
+        &self,
+        endpoint: &str,
+        discovery_handler_path: &str,
+    ) -> Option<StreamType> {
+        let discover_request = tonic::Request::new(DiscoverRequest {
+            discovery_details: self.config.spec.protocol.discovery_details.clone(),
+        });
+        trace!("get_stream - endpoint is {}", endpoint);
+        if endpoint == EMBEDDED_DISCOVERY_HANDLER_ENDPOINT
+            && get_discovery_handler(&self.config.spec.protocol.discovery_details).is_ok()
+        {
+            trace!(
+                "get_stream - internal discovery handler for protocol {}",
+                self.config.spec.protocol.name
+            );
+            let discovery_handler =
+                get_discovery_handler(&self.config.spec.protocol.discovery_details).unwrap();
+            Some(StreamType::Internal(
+                discovery_handler
+                    .discover(discover_request)
+                    .await
+                    .unwrap()
+                    .into_inner(),
+            ))
+        // Check if is a UDS connection
+        } else if endpoint.starts_with(discovery_handler_path) {
+            let path = endpoint.to_string();
+            match Endpoint::try_from("lttp://[::]:50051")
+                .unwrap()
+                .connect_with_connector(tower::service_fn(move |_: Uri| {
+                    let endpoint = path.clone();
+                    tokio::net::UnixStream::connect(endpoint)
+                }))
+                .await
+            {
+                Ok(channel) => {
+                    trace!(
```

Comment:
nit - Do we want to differentiate UDS vs other through in logging?

Reply:
fixed in 80e02b7.